pkg/server/delta/v3/server.go

```diff
@@ -44,6 +55,301 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
 	}
 }
 
+func (s *server) processDelta(str stream.DeltaStream, reqCh <-chan *discovery.DeltaDiscoveryRequest, defaultTypeURL string) error {
+	streamID := atomic.AddInt64(&s.streamCount, 1)
+
+	// unique nonce generator for req-resp pairs per xDS stream; the server
+	// ignores stale nonces. nonce is only modified within send() function.
+	var streamNonce int64
+
+	// a collection of stack allocated watches per request type
+	watches := NewWatches()
+
+	defer func() {
+		watches.Cancel()
+		if s.callbacks != nil {
+			s.callbacks.OnDeltaStreamClosed(streamID)
+		}
+	}()
+
+	// Sends a response, returns the new stream nonce
+	send := func(resp cache.DeltaResponse) (string, error) {
+		if resp == nil {
+			return "", errors.New("missing response")
+		}
+
+		out, err := resp.GetDeltaDiscoveryResponse()
+		if err != nil {
+			return "", err
+		}
+
+		streamNonce = streamNonce + 1
+		out.Nonce = strconv.FormatInt(streamNonce, 10)
+		if s.callbacks != nil {
+			s.callbacks.OnStreamDeltaResponse(streamID, resp.GetDeltaRequest(), out)
+		}
+
+		return out.Nonce, str.Send(out)
+	}
+
+	// Processes a response and updates the current server state
+	process := func(resp cache.DeltaResponse, more bool) error {
+		if resp == nil {
+			return errors.New("missing response")
+		}
+
+		typ := resp.GetDeltaRequest().GetTypeUrl()
+
+		if more {
+			watches.mu.Lock()
+			defer watches.mu.Unlock()
+
+			if resp == deltaErrorResponse {
```

Comment:
maybe have a helper `func isDeltaErrorResponse` to make this a little bit easier to grok?

Reply:
Nice I like this